tko-subs.go

```diff
@@ -170,273 +224,149 @@ func githubcreate(domain string) (bool, error) {
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}
 
-	//Now, for each entry in the data providers file, we will check to see if the output
-	//from the dig command against the current domain matches the CNAME for that data provider
-	//if it matches the CNAME, we need to now check if it matches the string for that data provider
-	//So, we curl it and see if it matches. At this point, we know its vulnerable
-	//Next, if we have the -takeover flag set to 1, we will also try to take over the dangling subdomain
+	client := &http.Client{Transport: transport, Timeout: time.Duration(10 * time.Second)}
+	var scanResults []ScanResult
 
 	for _, cmsRecord := range cmsRecords {
-
-		//by default, we try to curl using https but there are some CMS that take http only
-		//we specify this in the data providers file
-		protocol := "https://"
-		if cmsRecord.OverHTTP == "true" {
-			protocol = "http://"
-		}
-
-		usesprovider, err := regexp.MatchString(cmsRecord.CName, cname) //matching for the CNAME
-		Checkiferr(err)
-
-		if usesprovider { //if it matches the CNAME, create an entry in the final output
-			tkr.Domain = domain
-			tkr.IsTakenOver = false
-			tkr.IsVulnerable = false
-			tkr.Provider = cmsRecord.Name
-			tkr.RespString = cmsRecord.String
-
-			//Heroku behaves slightly different. Even if there is a dead DNS record for Heroku
-			//it would not resolve using host and you can't curl the website unlike other CMS
-			//but you will find it using dig
-			//So, if there is a CNAME match for heroku and can't curl it, we will assume its vulnerable
-			//if its not heroku, we will try to curl and regex match the string obtained in the response with
-			//the string specified in the data providers file to see if its vulnerable or not
-
-			response, err := client.Get(protocol + domain)
-			if err != nil && cmsRecord.Name == "heroku" {
-				isVulnerable = true
-				tkr.RespString = "Can't CURL it but dig shows a dead DNS record"
-			} else if err != nil && cmsRecord.Name != "heroku" {
-				fmt.Println(err)
-				panic(err)
-			} else if err == nil {
-				text, err := ioutil.ReadAll(response.Body)
-				Checkiferr(err)
-
-				isVulnerable, err = regexp.MatchString(cmsRecord.String, string(text))
-				Checkiferr(err)
-			}
-
-			//We now know if its vulnerable or not.
-			if isVulnerable {
-				tkr.IsVulnerable = true
-
-				switch *takeOver { //we know its vulnerable now. depending upon the flag to take over or not, we go forward
-				case true:
-					takenOver, err := takeoversub(domain, cmsRecord.Name)
-					Checkiferr(err)
-					if takenOver { //if successfully taken over
-						tkr.IsTakenOver = true
-					}
+		usesprovider, _ := regexp.MatchString(cmsRecord.CName, cname)
+		if usesprovider {
+			scanResult := evaluateDomainProvider(domain, cname, cmsRecord, client)
+			if (*config.takeOver && scanResult.IsVulnerable) {
+				_, err := takeoversub(scanResult.Domain, scanResult.Provider, config)
+				if (err != nil) {
```

Comment:
Shouldn't we be checking for err == nil here?

Reply:
Or if we are checking for err != nil, then it means we have an error and the scanResult.IsTakenOver should be set to false, correct?

Reply:
yep, noticed it, fixing thanks